### DISCUSSION TITLE
How to get service info (in partcular service cert) without a transaction

### DISCUSSION THREAD

**Discussion by jeffa5** [microsoft/CCF#4324] [Q&A] [answered]
Currently I can only see using a transaction to get a view on the service map as a way to get the service info. It would be useful if we could get this without having to use a transaction, e.g. in initialisation logic.

Is there a way to get it without a transaction?

**Comment by eddyashton** [comment, marked as answer]
The other way of accessing state is to use a `Subsystem`. See `node_configuration_subsytem.h` for an example of one which exposes the node's startup configuration.

The service cert is particularly difficult though, because it is not produced immediately on startup. So it might be accessible in `open()`, but not in your app/registry's constructor. There's also situations where the service cert changes - primarily on restarts in disaster recovery, but also on explicit [cert validity renewals by members](https://microsoft.github.io/CCF/main/operations/certificates.html#service-certificate).

What's the use case for reading service info without a transaction? If it's executed outside of a transaction (eg - in an index), we should consider what value of the `ServiceInfo` we want to use, and potentially read+cache it from an earlier transaction.

**Reply by jeffa5** [reply]
Interesting, thanks. The motivation was to be able to use it in the new post-commit logic, or get it in the startup to use throughout (which seems less than ideal for when it changes). I mainly need the cert to derive a whole-cluster ID (that can change when the cert changes). For now I'm ok working around it.
